File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/course/CourseMapper.java
package kh.farrukh.sortmapperexamplejava.course;

import kh.farrukh.sortmapperexamplejava.course.model.CourseCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseResponseDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseUpdateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.TeacherMapper;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingConstants.ComponentModel;
import org.mapstruct.MappingTarget;

@Mapper(
    componentModel = ComponentModel.SPRING,
    uses = TeacherMapper.class
)
public interface CourseMapper {

  @Mapping(target = "teacherId", source = "teacher.id")
  @Mapping(target = "teacherFirstName", source = "teacher.firstName")
  @Mapping(target = "teacherLastName", source = "teacher.lastName")
  CourseResponseDTO toResponseDTO(CourseEntity entity);

  CourseDetailsResponseDTO toDetailsResponseDTO(CourseEntity entity);

  CourseEntity toEntity(CourseCreateRequestDTO requestDTO);

  void update(@MappingTarget CourseEntity entity, CourseUpdateRequestDTO requestDTO);

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/course/CourseMapper.kt
package kh.farrukh.sortmapperexample.course

import kh.farrukh.sortmapperexample.course.model.CourseCreateRequestDTO
import kh.farrukh.sortmapperexample.course.model.CourseDetailsResponseDTO
import kh.farrukh.sortmapperexample.course.model.CourseResponseDTO
import kh.farrukh.sortmapperexample.course.model.CourseUpdateRequestDTO
import kh.farrukh.sortmapperexample.teacher.TeacherMapper
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.MappingConstants.ComponentModel
import org.mapstruct.MappingTarget

@Mapper(
    componentModel = ComponentModel.SPRING,
    uses = [TeacherMapper::class]
)
interface CourseMapper {

    @Mapping(target = "teacherId", source = "teacher.id")
    @Mapping(target = "teacherFirstName", source = "teacher.firstName")
    @Mapping(target = "teacherLastName", source = "teacher.lastName")
    fun toResponseDTO(entity: CourseEntity): CourseResponseDTO

    @Mapping(target = "teacherId", source = "teacher.id")
    fun toDetailResponseDTO(entity: CourseEntity): CourseDetailsResponseDTO

    fun toEntity(requestDTO: CourseCreateRequestDTO): CourseEntity

    fun update(@MappingTarget entity: CourseEntity, requestDTO: CourseUpdateRequestDTO)

}